src/ImageSharp/Memory/Allocators/Internals/GCAwareConfigurableArrayPool{T}.cs

```diff
@@ -0,0 +1,504 @@
+            return Numerics.Log2(((uint)bufferSize - 1) | 15) - 3;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static int GetMaxSizeForBucket(int binIndex)
+        {
+            int maxSize = 16 << binIndex;
+            Debug.Assert(maxSize >= 0, "Should be greater than 0.");
+            return maxSize;
+        }
+
+        /// <summary>
+        /// Provides a thread-safe bucket containing buffers that can be Rent'd and Return'd.
+        /// </summary>
+        private sealed class Bucket
+        {
+            private readonly T[][] buffers;
+            private readonly int poolId;
+            private readonly int numberOfBuffers;
+
+            /// <summary>
+            /// Do not make this readonly; it's a mutable struct
+            /// </summary>
+            private SpinLock spinLock;
+            private int index;
+            private uint firstItemMS;
+
+            /// <summary>
+            /// Initializes a new instance of the <see cref="Bucket"/> class.
+            /// </summary>
+            /// <param name="bufferLength">The length of each buffer.</param>
+            /// <param name="numberOfBuffers">The number of buffers each of <paramref name="bufferLength"/>.</param>
+            /// <param name="poolId">The pool id.</param>
+            internal Bucket(int bufferLength, int numberOfBuffers, int poolId)
+            {
+                this.spinLock = new SpinLock(Debugger.IsAttached); // only enable thread tracking if debugger is attached; it adds non-trivial overheads to Enter/Exit
+                this.buffers = new T[numberOfBuffers][];
+                this.BufferLength = bufferLength;
+                this.numberOfBuffers = numberOfBuffers;
+                this.poolId = poolId;
+            }
+
+            /// <summary>Gets an ID for the bucket to use with events.</summary>
+            internal int Id => this.GetHashCode();
+
+            internal int BufferLength { get; }
+
+            internal T[] Rent()
+            {
+                T[][] buffers = this.buffers;
+                T[] buffer = null;
+
+                // While holding the lock, grab whatever is at the next available index and
+                // update the index.  We do as little work as possible while holding the spin
+                // lock to minimize contention with other threads.  The try/finally is
+                // necessary to properly handle thread aborts on platforms which have them.
+                bool lockTaken = false, allocateBuffer = false;
+                try
+                {
+                    this.spinLock.Enter(ref lockTaken);
+
+                    if (this.index < buffers.Length)
+                    {
+                        buffer = buffers[this.index];
+                        buffers[this.index++] = null;
+                        allocateBuffer = buffer == null;
+                    }
+                }
```

Comment:
Calculating if buffer was null can be done outside of lock scope: `if(buffer == null) {}`
```suggestion
                bool lockTaken = false;
                try
                {
                    this.spinLock.Enter(ref lockTaken);

                    if (this.index < buffers.Length)
                    {
                        buffer = buffers[this.index];
                        buffers[this.index++] = null;
                    }
                }
```

Reply:
That's a straight copy from the runtime 

https://github.com/dotnet/runtime/blob/a02b668214e9711cb3d5f426f085601698a24b92/src/libraries/System.Private.CoreLib/src/System/Buffers/ConfigurableArrayPool.cs#L198-L217

Reply:
Note: I don't know if you can suggest changes from different places so **do not** commit this, line 353 must be changed  to `buffer == null` for this to work.

Reply:
@JimBobSquarePants hm, I may be wrong of course but we can easily do that outside of the lock scope, no? That comparison is nothing but with spinlocks better be fast than asleep :P